File: Leetcode Easy/110BalancedBinaryTree.py
#https://leetcode.com/problems/balanced-binary-tree/description/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        ret = True
        def helper(node):
            if node == None: return 0
            left = helper(node.left)
            right = helper(node.right)
            nonlocal ret
            if abs(left-right) > 1: ret = False
            return max(left, right) + 1
        helper(root)
        return ret

File: C++/Leetcode Medium/110BalancedBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <algorithm>
#include <cstdlib>
using namespace std;
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if (!root) return true;
        return helper(root).second;
    }
    pair<int, bool> helper(TreeNode* node) {
        if (!node) return {0, true};
        if (!node->left && !node->right) return {1, true};
        pair<int, bool> leftSubTree = helper(node->left);
        pair<int, bool> rightSubTree = helper(node->right);
        bool balanced = (leftSubTree.second && rightSubTree.second) && (abs(leftSubTree.first-rightSubTree.first)<=1);
        return {max(leftSubTree.first, rightSubTree.first)+1, balanced};
    }
};